src/runtime/cxx_unwind.cpp

```diff
@@ -0,0 +1,728 @@
+namespace __cxxabiv1 {
+
+extern "C" void* __cxa_allocate_exception(size_t size) noexcept {
+    // we should only ever be throwing ExcInfos
+    ASSERT(size == sizeof(pyston::ExcInfo), "allocating exception whose size doesn't match ExcInfo");
+
+    // Instead of allocating memory for this exception, we return a pointer to a pre-allocated thread-local variable.
+    //
+    // This variable, pyston::exception_ferry, is used only while we are unwinding, and should not be used outside of
+    // the unwinder. Since it's a thread-local variable, we *cannot* throw any exceptions while it is live, otherwise we
+    // would clobber it and forget our old exception.
+    //
+    // Q: Why can't we just use cur_thread_state.curexc_{type,value,traceback}?
+    //
+    // A: Because that conflates the space used to store exceptions during C++ unwinding with the space used to store
+    // them during C-API return-code based unwinding! This actually comes up in practice - the original version *did*
+    // use curexc_{type,value,traceback}, and it had a bug.
+    //
+    // In particular, we need to unset the C API exception at an appropriate point so as not to make C-API functions
+    // *think* an exception is being thrown when one isn't. The natural place is __cxa_begin_catch, BUT we need some way
+    // to communicate the exception info to the inside of the catch block - and all we get is the return value of
+    // __cxa_begin_catch, which is a single pointer, when we need three!
+    //
+    // You might think we could get away with only unsetting the C-API information in __cxa_end_catch, but you'd be
+    // wrong! Firstly, this would prohibit calling C-API functions inside a catch-block. Secondly, __cxa_end_catch is
+    // always called when leaving a catch block, even if we're leaving it by re-raising the exception. So if we store
+    // our exception info in curexc_*, and then unset these in __cxa_end_catch, then we'll wipe our exception info
+    // during unwinding!
+
+    return (void*)&pyston::exception_ferry;
+}
+
+// Takes the value that resume() sent us in RAX, and returns a pointer to the exception object actually thrown. In our
+// case, these are the same, and should always be &pyston::exception_ferry.
+extern "C" void* __cxa_begin_catch(void* exc_obj_in) noexcept {
+    assert(exc_obj_in);
+    pyston::us_unwind_resume_catch.log(pyston::per_thread_resume_catch_timer.end());
+
+    if (VERBOSITY("cxx_unwind"))
+        printf("***** __cxa_begin_catch() *****\n");
+
+    pyston::ExcData* e = (pyston::ExcData*)exc_obj_in;
+    e->check();
+    return (void*)&e->exc;
+}
+
+extern "C" void __cxa_end_catch() {
+    if (VERBOSITY("cxx_unwind"))
+        printf("***** __cxa_end_catch() *****\n");
+    // See comment in __cxa_begin_catch for why we don't clear the exception ferry here.
+}
+
+extern "C" void __cxa_throw(void* exc_obj, std::type_info* tinfo, void (*dtor)(void*)) {
+    assert(!pyston::in_cleanup_code);
+    assert(exc_obj);
```

Comment:
Unfortunately, I don't think so? The natural way to do it would be

```c++
RELEASE_ASSERT(tinfo == typeid(ExcInfo), "blah");
```

But `typeid` is disabled if we disable RTTI, which we do.